### PR TITLE
Send data in batches

### PR DESCRIPTION
**Pull request by patrobinson**
While using this plugin in production I discovered performance was extremely bad when the volume of events was more than a dozen per second. Output plugins are, by default, single threaded and since the `multi_receive` method isn't implemented it's just a single thread sending one message at a time to firehose. That basically limits this plugin to 10-20 events per second.

Firehose supports batch sends, which allows sending up to 500 at a time. This PR implements the `multi_receive` method, which Logstash automatically takes advantage of, to send up to 500 records at a time. We also check to ensure other limits, such as the maximum size of a single payload or single record, are not exceeded. Finally we refactor the tests to provide better guarantees it's working as required.

### CODE REVIEW

File: lib/logstash/outputs/firehose.rb [hunk 1]
```diff
@@ -43,12 +43,19 @@ class LogStash::Outputs::Firehose < LogStash::Outputs::Base
   TEMPFILE_EXTENSION = "txt"
   FIREHOSE_STREAM_VALID_CHARACTERS = /[\w\-]/
 
+  # These are hard limits
+  FIREHOSE_PUT_BATCH_SIZE_LIMIT = 4_000_000 # 4MB
```
Comment (chupakabr) [inline lib/logstash/outputs/firehose.rb:47]: would be nice to have these properties configurable

**Review by chupakabr — COMMENTED** [review]
thanks for your PR mate. Just curious, does logstash have an official repo of this plugin? I think I made PR to their repo long time ago..

File: lib/logstash/outputs/firehose.rb (hunk 1 above)
Comment (patrobinson) [reply lib/logstash/outputs/firehose.rb:47]: This is a hard limit from AWS, you cannot change it so I'm not sure the value in making it configurable.

**Comment by patrobinson** [comment]
@chupakabr there is, but your original PR has not been merged https://github.com/logstash-plugins/logstash-output-firehose

I figured this is as close to the upstream as I can get my changes.

Alternatively I can release this as a gem on Rubygems and become the unofficial maintainer

**Comment by chupakabr** [comment]
@patrobinson if you are fancy becoming an unofficial maintainer - feel free! :) I'll approve your PRs meanwhile here.
Cheers!

**Comment by chupakabr** [comment]
@patrobinson please resolve merge conflicts first, happened after merging your first PR ;)

**Comment by chupakabr** [comment]
cheers mate! @patrobinson

**Pull request merged**